src/Engine/EngineBase.php

```diff
@@ -17,29 +18,126 @@ abstract class EngineBase
     /** @var Intuition */
     protected $intuition;
 
-    public function __construct(Intuition $intuition)
+    /** @var KernelInterface */
+    protected $kernel;
+
+    /** @var string[][] Local PHP array copy of langs.json */
+    protected $langList;
+
+    /** @var string[] Additional localized names for non-standard language codes. */
+    public const LANG_NAMES = [
+        'az-cyrl' => 'Azərbaycan (qədim yazı)',
+        'de-frk' => 'Deutsch (Fraktur)',
+        'enm' => 'Middle English (1100-1500)',
+        'es-old' => 'español (viejo)',
+        'equ' => 'Math / equation detection module',
+        'fro' => 'Franceis, François, Romanz (1400-1600)',
+        'it-old' => 'italiano (vecchio)',
+        'ka-old' => 'ქართული (ძველი)',
+        'ko-vert' => '한국어 (세로)',
+        'osd' => 'Orientation and script detection module',
+        'sr-latn' => 'Српски (латиница)',
+        'syr' => 'leššānā Suryāyā',
+        'uz-cyrl' => 'oʻzbekcha',
+    ];
+
+    /**
+     * EngineBase constructor.
+     * @param Intuition $intuition
+     * @param KernelInterface $kernel
+     */
+    public function __construct(Intuition $intuition, KernelInterface $kernel)
     {
         $this->intuition = $intuition;
+        $this->kernel = $kernel;
     }
 
     /**
+     * Unique identifier for the engine.
+     * @return string
+     */
+    abstract public static function getId(): string;
+
+    /**
+     * Get transcribed text from the given image.
      * @param string $imageUrl
      * @param string[]|null $langs
      * @return string
      */
     abstract public function getText(string $imageUrl, ?array $langs = null): string;
 
     /**
+     * Get the language list from langs.json
+     * @return string[][]
+     */
+    private function getLangList(): array
+    {
+        if (!$this->langList) {
+            $this->langList = json_decode(file_get_contents($this->kernel->getProjectDir().'/public/langs.json'), true);
```

Comment:
This is the only use of $kernel isn't it? Would it be better to inject the project directory only, rather than the whole kernel? e.g. in services.yaml:

```lang=yaml
    App\Engine\EngineBase:
        arguments:
            $projectDir: '%kernel.project_dir%'
```

Reply:
Good call! Done.